lib/spack/env/cc

```diff
@@ -412,6 +412,22 @@ esac
 # Linker flags
 case "$mode" in
     ld|ccld)
+        # This is needed to ensure we set RPATH instead of RUNPATH.
+        # Documentation on this mechanism is lacking at best. A few sources
+        # of information are (note that some of them take explicitly the
+        # opposite stance that Spack does):
+        #
+        # http://blog.qt.io/blog/2011/10/28/rpath-and-runpath/
+        # https://wiki.debian.org/RpathIssue
+        #
+        # The only discussion I could find on enabling new dynamic tags by
+        # default on ld is the following:
+        #
+        # https://sourceware.org/ml/binutils/2013-01/msg00307.html
+        #
+        flags+=('-Wl,--disable-new-dtags')
```

Comment:
This seems to me the cleanest fix. However I don't have access to Cray machines or to PGI or IBM compilers so I don't know if this is generic enough. Can you guys check if I need to do something more than this (like set some environment variable that will then be expanded in the wrapper)?

Reply:
I doubt this will work for NAG, which doesn't even support `-Wl,-rpath`.

Reply:
`ld` and `ccld` mode are different -- one's the linker and the other's the compiler.  In linker mode, you'd just pass `--disable-new-dtags`.  But that's not sufficient b/c that's only an argument for GNU `ld`, `gold`, and other linkers that handle ELF.

On a mac (which uses mach-o instead of ELF and doesn't have this elf-specific argument), we should skip the argument entirely.

Per @adamjstewart's comment -- I think the linker argument prefix should be passed in.  We currently do that with `SPACK_<lang>_RPATH_ARG`.

I can't decide whether there should be a generic prefix variable like `SPACK_LINKER_ARG_PREFIX` for `ccld` mode, in which case the compiler would know about `--disable-new-dtags` and `--enable-new-dtags`, or whether there should be some other more specific thing like `SPACK_ENABLE_DTAGS` and `SPACK_DISABLE_DTAGS`.  Either way there needs to be some support from the compiler classes.

Reply:
@adamjstewart @tgamblin Can you check if this implementation fits NAG and non-linux systems better? If it does I'll try to add other tests to leverage those cases.

> The other thing that should be handled here is if the build provides --enable-new-dtags. We need to strip that, because I dunno what happens if both --enable-new-dtags and --disable-new-dtags are provided together on the link line. We want disable to be all that is left.

This still needs to be implemented.